### PR TITLE
Clean up and clarify Normalize docs

### PR DESCRIPTION
**Pull request by dstansby**
I'm trying to debug `SymLogNorm`, so thought I'd start by tidying up the `Normalize` docs to make them clearer. Suggestions very welcome.

### CODE REVIEW

File: lib/matplotlib/colors.py [hunk 1]
```diff
@@ -920,29 +920,59 @@ def reversed(self, name=None):
...
+    with the signatures given below.
     """
-    A class which, when called, can normalize data into
-    the ``[0.0, 1.0]`` interval.
+    @abc.abstractmethod
+    def __call__(self, value, clip=None):
+        """
+        Normalize *value* into the range ``[0, 1]``.
+        """
+        ...
```
Comment (anntzer) [inline lib/matplotlib/colors.py]: You don't need the ... (the docstring is sufficient as a function body).

File: lib/matplotlib/colors.py [hunk 2]
```diff
@@ -955,15 +985,19 @@ def process_value(value):
...
 
-        Returns *result*, *is_scalar*, where *result* is a
-        masked array matching *value*.  Float dtypes are preserved;
-        integer types with two bytes or smaller are converted to
-        np.float32, and larger types are converted to np.float64.
-        Preserving float32 when possible, and using in-place operations,
-        can greatly improve speed for large arrays.
+        Returns
+        -------
+        result : maksed array
```
Comment (anntzer) [inline lib/matplotlib/colors.py]: typo: masked

File: lib/matplotlib/colors.py [hunk 3]
```diff
@@ -955,15 +985,19 @@ def process_value(value):
...
+            Masked array matching *value*.
```
Comment (anntzer) [inline lib/matplotlib/colors.py]: matching -> with value's shape / with the same shape as value (e.g.)

**Comment by anntzer** [comment]
I agree with splitting an "abstract" normbase from the linear norm.

File: lib/matplotlib/colors.py [hunk 4]
```diff
@@ -981,11 +1013,21 @@ def process_value(value):
...
-        the ``[0.0, 1.0]`` interval and return it.  *clip* defaults
-        to ``self.clip`` (which defaults to *False*).  If not already
-        initialized, *vmin* and *vmax* are initialized using
-        ``autoscale_None(value)``.
+        Normalize *value* data in the ``[vmin, vmax]`` interval into the
+        ``[0.0, 1.0]`` interval and return it.
+
+        Parameters
+        ----------
+        value :
```
Comment (anntzer) [inline lib/matplotlib/colors.py]: no colon

**Comment by timhoffm** [comment]
The naming `Normalize` has been annoying me for a while. It breaks with every other norm naming, which is `XxxNorm` and it's really imprcise. I would be very much in favor of soft-deprecating it in favor of `LinearNorm`. `Normalize can stay for back-compatibility, but not actively using it in modern code would increase readability and consistency quite a bit.

I agree on adding an abstract norm class. Not a fan of naming it `NormBase`. I recognize that we name all base classes `XxxBase`, which feels not quite right - neither the word order nor using `Base` at all. Even though it would be moving away from an estrablished pattern, could we use `AbstractNorm`?

**Comment by dstansby** [comment]
👍 sounds good, I'll take a go at that in this PR. How about just `Norm`? Maybe too simple.

**Comment by anntzer** [comment]
I guess BaseNorm would be somewhat more idiomatic but just weird given all the XxxBase we have.  I think the problem of AbstractNorm is that it's also what you'd use in all docstrings which currently read `norm : Normalize`; `norm : AbstractNorm` is just weird whereas `norm : BaseNorm` or `norm : NormBase` seem OK?

**Comment by dstansby** [comment]
I'm going to leave this PR as just doc changes, and put the other stuff elswehere since it's turning out to be a bit of a big diff.

**Comment by anntzer** [comment]
failure is the scipy one.

**Pull request merged**